client/data/marketplace/site-search-api.ts

```diff
@@ -0,0 +1,96 @@
+import { flatten } from 'q-flat';
+import { encode } from 'qss';
+import { RETURNABLE_FIELDS } from './constants';
+import type { SearchParams } from './types';
+
+// Maps sort values to values expected by the API
+const SORT_QUERY_MAP = new Map( [
+	[ 'oldest', 'date_asc' ],
+	[ 'newest', 'date_desc' ],
+	[ 'relevance', 'score_default' ],
+] );
+/**
+ * Map sort values to ones compatible with the API.
+ *
+ * @param {string} sort - Sort value.
+ * @returns {string} Mapped sort value.
+ */
+function mapSortToApiValue( sort: string ) {
+	// Some sorts don't need to be mapped
+	if ( [ 'price_asc', 'price_desc', 'rating_desc' ].includes( sort ) ) {
+		return sort;
+	}
+
+	return SORT_QUERY_MAP.get( sort ) ?? 'score_default';
+}
+
+/**
+ * Generate the query string for an API request
+ *
+ * @returns {string} The generated query string.
+ */
+function generateApiQueryString( { query, groupId, pageHandle, pageSize, locale }: SearchParams ) {
+	const sort = 'score_default';
+
+	let params: {
+		fields: string[];
+		filter?: { bool: { must: object[] } };
+		page_handle?: string;
+		query: string;
+		sort: string;
+		size: number;
+		group_id?: string;
+		from?: number;
+		lang: string;
+	} = {
+		fields: [ ...RETURNABLE_FIELDS ],
+		// filter: buildFilterObject( filter ),
+		page_handle: pageHandle,
+		query: encodeURIComponent( query ?? '' ),
+		sort: mapSortToApiValue( sort ),
+		size: pageSize,
+		lang: locale,
+	};
+
+	if ( groupId ) {
+		params = {
+			...params,
+			group_id: groupId,
+		};
+	}
+
+	// TODO implement author search
+	// if ( author ) {
+	// }
+
+	return encode( flatten( params ) );
+}
+
+/**
+ * Perform a search.
+ *
+ * @param {object} options - Search options
+ * @returns {Promise} A promise to the JSON response object
+ */
+export function search( options: SearchParams ) {
+	const queryString = generateApiQueryString( options );
+
+	const pathForPublicApi = `/marketplace/search?${ queryString }`;
+
+	const url = `https://public-api.wordpress.com/rest/v1.3${ pathForPublicApi }`;
+
```

Comment:
Should we move these to constants?

Reply:
Do you think we should try to use here the `wpcom.req` client like [this example](https://github.com/Automattic/wp-calypso/blob/0291f808682499c108cd311e4fa7430ab8815313/client/data/marketplace/use-wpcom-plugins-query.ts#L63-L64)? I am working on adding it but currently finding issues that I should be able to sort out.

Reply:
Yes, it would be great. I remember @gcsecsey also having problems due to different structure of the response when he was directly requesting from the jetpack es endpoint. But now that we have create a new endpoint, we should be able to format the response in the backend so that we can use `wpcom.req`. Can you try using `return rest_ensure_response($response);` in the `callback` of `WPCOM_JSON_API_Marketplace_Search_Endpoint`?